### PR TITLE
Remove int2type and bool2type

### PR DESCRIPTION
**Pull request by masterleinad**
Using C++11 we can use std::integral_constant instead.

### CODE REVIEW

**Comment by tjhei** [comment]
Nice! :+1:  /run-tests

Are we okay with outright removing it because it was "internal"?

**Comment by bangerth** [comment]
Looks good if you bring back the class and `DEAL_II_DEPRECATED` it.

**Comment by drwells** [comment]
For the record (to record my comments here): I would prefer to keep `int_2_type` and instead slowly remove usages of `int_2_type` from the library since we can now partially specialize template functions.

**Comment by masterleinad** [comment]
@bangerth The classes are back again (and deprecated).

**Comment by bangerth** [comment]
OK to merge once the tester is happy.

File: include/deal.II/base/template_constraints.h [hunk 1]
```diff
@@ -254,21 +254,25 @@ namespace internal
    * still select which of the different <code>X::f()</code> we want based on
    * the <code>subdim</code> template argument.
    *
+   * @deprecated Use std::integral_constant instead.
```
Comment (tjhei) [inline include/deal.II/base/template_constraints.h]: can you say ``std::integral_constant<int,N>``?

File: include/deal.II/base/template_constraints.h [hunk 2]
```diff
@@ -254,21 +254,25 @@ namespace internal
...
   template <int N>
   struct int2type
-  {};
+  {} DEAL_II_DEPRECATED;
 
 
   /**
    * The equivalent of the int2type class for boolean arguments.
    *
+   * @deprecated Use std::integral_constant instead.
```
Comment (tjhei) [inline include/deal.II/base/template_constraints.h]: ``<bool, N>``

**Comment by tamiko** [comment]
We have a smallish fallout of ~7 tests reported by the tester due to missing instantiations.

**Comment by masterleinad** [comment]
I'm on it.

**Comment by masterleinad** [comment]
This should work now. Locally, all `matrix_free` tests are now passing for me.

**Pull request merged**